### PR TITLE
Use contourpy for quad contour calculations

### LINKED ISSUE #22529: Replace C++ quad contouring code with use of ContourPy
I propose to remove the C++ quad contouring code that is in Matplotlib, which I originally wrote in 2014, and replace it with use of a new external dependency [ContourPy](https://github.com/contourpy/contourpy).

ContourPy is a new Python/C++ project that contains the original (2005) and current (2014) Matplotlib contouring algorithms as well as a new one that includes more features and is faster is many situations. The project primarily exists to allow other Python libraries to have access to contouring without having to have Matplotlib as a dependency, but I am happy to do the work required for it to be used by Matplotlib as well.

The new algorithm is a complete rewrite, taking the good features of the previous two algorithms and replacing the bad features with better ones. There are new features which may be of interest eventually, such as the `quad_as_tri` option which divides each full quad into 4 triangles for contouring purposes, and support for multithreading (high risk at the moment). The major difference is that it supports a number of different return types so that, for example, you can request Matplotlib `Path` kind codes or not.

I recommend a conservative approach, which is to use the new algorithm but without any of the new features and using exactly the same return types that Matplotlib is used to. Later on we can look at those two areas which will give improved functionality and performance.

The changes for this can be done in a single PR (which I have waiting in the wings) as follows:
1. Delete the three `src/_contour*` files and remove the `contour` section from `setupext.py`.
2. Add the new dependency `contour>=1` to `setup.py`, `environment.yml` and possibly other places in the CI config files.
3. In `contour.py` change the import statement and creation of a `ContourGenerator` object. This changes only ~6 lines of code.
4. Removing the tests of the internal contour API in `test_contour.py`.
5. Baseline image changes.

The latter all occur because the new contouring algorithm returns exactly the same lines/polygons but not necessarily starting with the same point. The Agg backend renders polygons with 4+ points exactly the same regardless of starting point, but it renders triangles differently. This is a feature/bug of Agg which has been around for years.

Your concerns should be ones of packaging and maintenance.

## Packaging

ContourPy is written in C++11 and wrapped using [pybind11](https://github.com/pybind/pybind11). I don't use any C++ language features invented in the last 10+ years. Pybind11 is actively maintained and builds on all hardware/OS/Python version combinations. I have [PyPI wheels](https://pypi.org/project/contourpy/#files) available for all the hardware/OS/Python version combinations that Matplotlib supports plus a few others, all via `cibuildwheel`. There are [conda-forge](https://anaconda.org/conda-forge/contourpy) packages too.

The only runtime dependency is `numpy`.

## Maintenance

I am the only maintainer of ContourPy and I expect it will stay that way although obviously I would welcome contributions from others. My job (https://makepath.com/) is in the areas of visualisation and spatial algorithms and ContourPy sits firmly in both camps, and I have a remit to use, develop and support Open Source projects, so the short to medium term future of ContourPy is assured. I suppose the other comment here is that I am already the de facto maintainer of the Matplotlib C++ contouring and tricontouring algorithms that I wrote, but I can only recall fixing one bug in each over the last 8 and 12 years respectively, both just after initial release, so you should not have any particular concerns about the quality of my C++.

If you don't want to go ahead with this change and would prefer to keep the existing C++ code in Matplotlib then I am fine with this. I am also OK with the idea of using ContourPy but via the existing `mpl2014` algorithm (as I call it) which involves a PR with all the changes above except for the baseline image changes.

### PR DESCRIPTION
**Pull request by ianthomas23**
Closes #22529.

This PR replaces the C++ contouring code shipped with Matplotlib with a new external dependency `contourpy`. Details are listed in the linked issue.

Nearly all of the test image changes are too small for the human eye to identify and are caused by a change in the start point of calculated and rendered polygons. Other test image changes are visible and are all movements of contour labels, code that has not changed in this PR but evidently the change in order of the lines/polygons returned affects the `clabel` positions.

I have added `contourpy` as a dependency in `setup.py` and `environment.yml`. I don't think there are any other files featuring dependencies that need to be changed here, but I am not 100% sure.

### CODE REVIEW

**Comment by anntzer** [comment]
I would suggest starting with sticking to the mpl2014 algorithm for now so that we don't have to update the images?  Actually switching the algorithm (perhaps configurable globally via an rcParam?) could be done as a second step?

**Comment by anntzer** [comment]
Do you know what happened to patheffect2?  The change isn't a huge deal, but the old label position seems better?

**Comment by ianthomas23** [comment]
> Do you know what happened to patheffect2? The change isn't a huge deal, but the old "18" label position seems better?

I do, and it is most unfortunate.

The `test_patheffects2` images are different as it uses `contour` to generate some lines to put the path effects on. There is a change I purposefully put in contourpy's mpl2014 algorithm that is not in mpl's version of it. Mpl's doesn't allow identical adjacent (x,y) points in a contour line/polygon, but contourpy does. This is because some rendering/processing libraries that consume polygon data very sensibly only allow polygons that contain at least 3 points. So the 2 versions produce the same visual output in Mpl but the underlying point arrays from contourpy may be longer with adjacent duplicate points.

Mpl's label positioning algorithm evidently uses the number of points in a line to work out where to put the label, so it can be different now.  If the contour lines weren't going exactly through the grid points there would be no difference.

**Comment by ianthomas23** [comment]
The minimum versions CI test is failing because in contourpy I have pinned a minimum numpy version of 1.20 whereas it is 1.19 in mpl. It doesn't really need to be 1.20 so I can put it back to 1.19, but that will need a new release of contourpy.

**Comment by ianthomas23** [comment]
> Actually switching the algorithm (perhaps configurable globally via an rcParam?) could be done as a second step?

I am not sure about the idea of using a new rcParam. In the long run we will use the new `serial` algorithm as that is the only way we will get the improvements in functionality and performance. When that happens, do we deprecate the new rcParam or add documentation to explain the available options? Maybe it would be better to make the decision for end users so that they don't have to learn about the alternatives. But I happy to be guided by others on this.

**Comment by QuLogic** [comment]
We have lots of machinery for deprecating rcParams, cf. #13761, #15318, #17156, but I don't know whether it's the best choice to use one here.

**Comment by ianthomas23** [comment]
We probably need to agree a strategy going forward here. Do we run with this as it is (current algorithm) for one mpl release and then switch to the new algorithm for the next mpl release to spread the risk, or do we do both changes in one go?

**Comment by ianthomas23** [comment]
I wish you had given me some notice that you were going to talk about this on the call so that I could attend and give an informed opinion. The algorithms already have names (https://contourpy.readthedocs.io/en/v1.0.1/user_guide/name.html) and the names in question here are `"mpl2014"` and `"serial"`. Obviously they wouldn't be called `"old"` and `"new"`.

I still need some clarity here. Sticking with `"mpl2014"` for Matplotlib 3.6 is fine, in which case you don't need a kwarg until after 3.6. Or do you mean you'd like a kwarg now but the default would remain as `"mpl2014"`? The logical progression from this is that you need to include some tests for both possible kwargs (or indeed all 4 that ContourPy supports) and hence you are going to have to include two (or possibly 4?!) sets of baseline images.

**Comment by tacaswell** [comment]
@ianthomas23 Sorry about that, will you be able to make the call next week?

The analogy we had was the interpolation kwarg on `imshow`.  My position is that we should not block this PR on getting a kwarg is place for 3.6.  If we get a kwarg (that defaults to `'mpl2014'`) on for 3.6 👍🏻 , but we should also bank the win of depending on contourpy as-is!

If there are 4 algorithms I would go with something like https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/tests/baseline_images/test_image/imshow_masked_interpolation.png to make sure they all run and leave a majority of our test images using `'mpl2014'`.   We might even be able to reduce some of our contour tests as I am happy to trust that the testing on contourpy is doing a more complete job of testing the algorithms than we would want to do.

**Comment by ianthomas23** [comment]
> @ianthomas23 Sorry about that, will you be able to make the call next week?

Yes, I can make this week's call.

Thanks for the clarification, it makes more sense now. Getting this all done for 3.6 should be fine. To do list (mostly for my own benefit):

1. Add kwarg for algorithm name. I am using e.g. `name='mpl2014'` in contourpy. For `mpl` using `name=...` is maybe not precise enough but the clearer `algorithm=...` or `algo=...` are possibly worse choices. Maybe `name=...` is OK if the docs refer to 'algorithm name'.
2. If algorithm name is not specified by the user, it can read an `rcparam` just like `contour.corner_mask`, which will initially be `'mpl2014`'.
3. Probably should validate the `name` in mpl against the list of acceptable strings. This already happens in contourpy of course but we should protect against the possibility of dodgy installs combining old and new version of contourpy/mpl.
4. We'll allow `name='mpl2005'` (the original non-corner mask algorithm) and `name='threaded'` (experimental multithreaded version of `'serial'`). There shouldn't be many users of either but the former can be used by people who have been using the `legacycontour` repo (for whatever reason?) and the latter will improve and become a reasonable option eventually.
5. Keep all of the existing tests as they are using `'mpl2014'`, and add one or two new ones to test the different algorithms. This is mostly to test that mpl and contourpy are talking to each other correctly, comprehensive testing off all algorithms occurs in contourpy.
6. Update docs, change note and API change for new kwarg. The docs will have to list the valid kwarg values but should avoid too much description and link to contourpy docs instead.
7. Doesn't need a new example, there is effectively no change for end users who are leaving the kwarg alone.

Once this is in 3.6 and we have confirmed everything is OK, a future PR will:

1. Switch the default to `'serial`', via the rcparam.
2. Update the docs to explain the change.
3. Update the contour test baseline images to use `'serial'` not `'mpl2014'`, and remove some of the tests as most of that responsibility is devolved to contourpy.

After that, there is the possibility of taking advantage of the improvements in `'serial'`, i.e.

1. Performance improvements by returning contour data in slightly different format from contourpy.
2. New functionality such as `quad_as_tri` option and different `z_interp` options.

I am keen on doing item 1. Item 2 might be driven by user requests for such.

**Comment by ianthomas23** [comment]
Updates following yesterday's call:

1. Keyword argument will be called `algorithm`.
2. There is a new release of contourpy (v1.0.2) as requested by @QuLogic following the changes to the tests.

**Comment by QuLogic** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=2073692; no issues with tests on other architectures.

File: lib/matplotlib/mpl-data/matplotlibrc [hunk 1]
```diff
@@ -607,10 +607,11 @@
 ## * CONTOUR PLOTS                                                           *
 ## ***************************************************************************
 #contour.negative_linestyle: dashed  # string or on-off ink sequence
-#contour.corner_mask:        True    # {True, False, legacy}
```
Comment (jklymak) [inline lib/matplotlib/mpl-data/matplotlibrc:610]: This probably needs to be in the API change as well?

File: doc/users/next_whats_new/use_contourpy.rst [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+.. note::
+
+   Contour lines and polygons produced by ``algorithm='mpl2014'`` will be the
+   same as those produced before this change to within floating-point
+   tolerance. The exception is for duplicate points, i.e. contours containing
+   adjacent (x, y) points that are identical; previously the duplicate points
+   were removed, now they are kept. Contours affected by this will produce the
+   same visual output, but there will be a greater number of points in the
+   contours.
```
Comment (jklymak) [inline doc/users/next_whats_new/use_contourpy.rst:24]: Do the differences in clabel need to be mentioned here or are those usually small enough not to matter?

**Review by jklymak — COMMENTED** [review]
Seems good to me, but maybe a few more things need to be mentioned in the changes notes?

File: lib/matplotlib/mpl-data/matplotlibrc (hunk 1 above)
Comment (ianthomas23) [reply lib/matplotlib/mpl-data/matplotlibrc:610]: Support for `corner_mask='legacy'` was deprecated in mpl 1.5 (doc/api/prev_api_changes/api_changes_1.5.0.rst) and the code removed in mpl 2.2 (`doc/api/prev_api_changes/api_changes_2.2.0.rst`). This evidently missed the code removal.

File: doc/users/next_whats_new/use_contourpy.rst (hunk 2 above)
Comment (ianthomas23) [reply doc/users/next_whats_new/use_contourpy.rst:24]: Yes, I will mention it here.

**Review by jklymak — APPROVED** [review]
Looks good to me.  Did you want to keep all the commits, or should this be squashed?

**Comment by ianthomas23** [comment]
I'd squash it.

**Comment by ianthomas23** [comment]
Commits squashed.

**Review by QuLogic — APPROVED** [review]

**Comment by jklymak** [comment]
@ianthomas23 I squashed and merged your latest changes.  Thanks for doing this - hopefully this is a smooth transition!

**Pull request merged**